### PR TITLE
🐛 Fix `DioErrorType.cancel` in `Interceptors`

### LINKED ISSUE #1512: Interceptor doesn't work for DioError.type = cancel
Interceptor have no effect on dio error when you cancel request.

**version**: 4.0.6

**Example**

```
class ErrorInterceptor extends Interceptor {
  @override
  void onError(DioError err, ErrorInterceptorHandler handler) {
    if (err.type == DioErrorType.cancel) {
      handler.next(CancelledException());
    } else {
      handler.next(err);
    }
  }
}
```

When register interceptor and try to catch an error it always has a type = `DioError` instead of `CancelledException` whenever request is cancelled.

### PR DESCRIPTION
**Pull request by AlexV525**
Fixes #1512.

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### Additional context and info (if any)

Previously, `listenCancelForAsyncTask` will wait for `CancelToken.whenCancel` in certain scenarios:
https://github.com/cfug/dio/blob/6544701ed370e6ceaf81576f393459605e9a612f/dio/lib/src/dio_mixin.dart#L457-L466

These scenarios include `DioErrorType.cancel`, in which the error handler has been put into the event loop rather than executed immediately, which means the handler will handle the exception after the request is completed.

### CODE REVIEW

File: dio/test/mock/adapters.dart [hunk 1]
```diff
@@ -32,32 +32,30 @@ class MockAdapter implements HttpClientAdapter {
             },
           );
         case '/test-auth':
-          {
```
Comment (AlexV525) [inline dio/test/mock/adapters.dart:35]: A format change to remove the `Set` caller.

File: dio/test/mock/adapters.dart [hunk 2]
```diff
@@ -70,23 +68,23 @@ class MockAdapter implements HttpClientAdapter {
               },
             );
           });
-
         case '/token':
-          {
```
Comment (AlexV525) [inline dio/test/mock/adapters.dart:75]: Same as above.

File: dio/test/mock/adapters.dart [hunk 3]
```diff
@@ -70,23 +68,23 @@ class MockAdapter implements HttpClientAdapter {
...
+            200,
+            headers: {
+              Headers.contentTypeHeader: [Headers.jsonContentType],
+            },
+          );
         case '/test-force-convert':
           return ResponseBody.fromString('{"code":0,"result":"ok"}', 200);
+        case '/test-timeout':
+          await Future.delayed(const Duration(days: 365));
+          return ResponseBody.fromString('', 200);
```
Comment (AlexV525) [inline dio/test/mock/adapters.dart:87]: Adds a timeout handler that is intended to be timed out in the whole year.

File: dio/test/interceptor_test.dart [hunk 4]
```diff
@@ -469,6 +469,40 @@ void main() {
...
+      DioError? iError, qError;
+      final dio = Dio()
+        ..httpClientAdapter = MockAdapter()
+        ..options.baseUrl = MockAdapter.mockBase
+        ..interceptors.add(
+          InterceptorsWrapper(
+            onError: (DioError e, ErrorInterceptorHandler handler) {
+              iError = e;
+              handler.next(e);
+            },
```
Comment (AlexV525) [inline dio/test/interceptor_test.dart:484]: Actually, the caller will be called previously, but it'll be slightly after the expectation. If we add any delay after awaiting the request, the test will be passed.

**Review by chenglu — APPROVED** [review]

**Pull request merged**